tools/OMIcheck/OMIcheck.ps1

```diff
@@ -1,127 +1,460 @@
+        Write-Host -ForegroundColor Red `t`t $VM.Name ": VM has vulnerable OMI version " $check.OmiPkgVer
+        if ($upgradeOMI)
+        {
+            if ($check.LadPkgVer -ne $LadPkgGoodVersion)
             {
-                Write-Host -ForegroundColor Gray `t`t $VM.Name " in resource group : VM is not running"
-                continue
+                UpdateLAD($VM)
             }
 
-            # Update Linux only VMs.
-            if ($VM.StorageProfile.OsDisk.OsType.ToString() -ne "Linux")
+            if ($check.OmsPkgVer -ne $OmsPkgGoodVersion)
             {
-                Write-Host -ForegroundColor Gray `t`t $VM.Name ": VM is not running Linux OS"
-                continue
+                UpdateOMS($VM)
             }
 
-            # TODO: Consider setting timeout. Parameter does not exist, -AsJob is an option for v2.
-            $check = Invoke-AzVMRunCommand -ResourceGroupName $VM.ResourceGroupName -Name $VM.Name -CommandId 'RunShellScript' -ScriptPath $checkScriptPath
+            # Force upgrade OMI pkg just in case.
+            UpgradeVmOMI($VM)
+        }
+
+        # DEBUG:
+        #break
+    }
+}
+
+# Update the VM extensions and OMI package in a subscription for VM scale sets.
+# TODO: This can be converted to ForEach-Object -Parallel. However, it requires PSh >= 7.1
+function UpdateVMss($sub)
+{
+    # Set Azure Subscription context
+    Set-AzContext -Subscription $sub.Id
+
+    Write-Output "Listing Virutal Machine scale sets in subscription '$($sub.Name)'"
+    $VMsss = Get-AzVmss
+    $VMsssSorted = $VMsss | Sort-Object -Property ResourceGroupName
+    $PreviousRG = ""
+    foreach($VMss in $VMsss)
+    {
+        # DEBUG: limit to 1 vmss only
+        #if ($VMss.name -ne "my-testing-vmss")
+        #{
+        #    continue
+        #}
 
-            $split = $check.Value.Message.Split(" ",[System.StringSplitOptions]::RemoveEmptyEntries)
-            if ($check.Value.Message.Contains("/opt/omi/bin/omiserver:"))
+        # Group VMss by resource groups.
+        if ($PreviousRG -ne $VMss.ResourceGroupName)
+        {
+            Write-Host "`tResourceGroup: " $VMss.ResourceGroupName
+            $PreviousRG = $VMss.ResourceGroupName
+        }
+
+        # Update only running VMss.
+        if ($VMss.ProvisioningState -ne "Succeeded")
+        {
+            Write-Host -ForegroundColor Gray `t`t $VMss.Name " VMss is not in succeeded state: " $VMss.ProvisioningState
+            continue
+        }
+
+        $VMinstances = Get-AzVmssVM -ResourceGroupName $VMss.ResourceGroupName -VMScaleSetName $VMss.Name -InstanceView
+        if (!(IsVMRunningLinux($VMinstances[0].InstanceView)))
+        {
+            continue
+        }
+        if ($VMss.VirtualMachineProfile.OsProfile.LinuxConfiguration -eq $null)
+        {
+            Write-Host -ForegroundColor Gray `t`t $VMss.Name " is not running Linux. Skipping."
+            continue
+        }
+
+        $check = IsVMssVulnerableOMI $VMss $VMinstances[0]
```

Comment:
seems unlikely but in this case, if the first instance was different than other instances (e.g. for some reason doesn't have OMI installed but others do), we could skip when we should actually be patching

Reply:
VMSS is uniform; that's same bits are on all instances. We got to pick an instance. Going thru all of them would take long time.